src/bls12_381/ec.rs

```diff
@@ -130,18 +136,60 @@ macro_rules! curve_impl {
                     let mut y2 = self.y;
                     y2.square();
 
-                    let mut x3b = self.x;
-                    x3b.square();
-                    x3b.mul_assign(&self.x);
-                    x3b.add_assign(&Self::get_coeff_b());
-
-                    y2 == x3b
+                    y2 == y2_from_x(self.x)
                 }
             }
 
             fn is_in_correct_subgroup_assuming_on_curve(&self) -> bool {
                 self.mul($scalarfield::char()).is_zero()
             }
+
+            fn sw_encode(t: $basefield) -> Self {
+                // handle the case t == 0
+                if t.is_zero() { return Self::zero() };
+
+                // We choose the corresponding y-coordinate with the same parity as t.
+                let parity = t.parity();
+
+                // w = (t^2 + b + 1)^(-1) * sqrt(-3) * t
+                let mut w = t;
+                w.square();
+                w.add_assign(&get_coeff_b());
+                w.add_assign(&$basefield::one());
+                // handle the case t^2 + b + 1 == 0
+                if w.is_zero()  {
+                    let mut ret = Self::one();
+                    if !parity { ret.negate() }
+                    return ret
+                };
+                w = w.inverse().unwrap();
+                w.mul_assign(&$basefield::get_swenc_sqrt_neg_three());
+                w.mul_assign(&t);
+
+                // x1 = - wt  + (sqrt(-3) - 1) / 2
+                let mut x1 = w;
+                x1.mul_assign(&t);
+                x1.negate();
+                x1.add_assign(&$basefield::get_swenc_sqrt_neg_three_minus_one_div_two());
+                if let Some(p) = Self::get_point_from_x(x1, parity) {
+                    return p
+                }
+
+                // x2 = -1 - x1
+                let mut x2 = x1;
+                x2.negate();
+                x2.sub_assign(&$basefield::one());
+                if let Some(p) = Self::get_point_from_x(x2, parity) {
+                    return p
+                }
+
+                // x3 = 1/w^2 + 1
+                let mut x3 = w;
+                x3.square();
+                x3 = x3.inverse().unwrap();
+                x3.add_assign(&$basefield::one());
+                Self::get_point_from_x(x3, parity).unwrap()
```

Comment:
Instead of `.unwrap`, let's use `.expect("this point must be valid if the other two are not")`